primitives/treasury-funding/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+        let maximum_cut = T::MaximumCut::get();
+        let delta_cut = maximum_cut.saturating_sub(minimum_cut);
+
+        // Calculate the final treasury cut percentage by adjusting the minimum cut with the adjustment times the delta cut.
+        let treasury_cut: FixedU128 = minimum_cut.saturating_add(delta_cut * adjustment).into();
+
+        // Calculate the amount to transfer to the treasury by using the equation `amount_to_charge * treasury_cut_percentage`,
+        // where `treasury_cut_percentage` is obtained by getting the inner value of `treasury_cut` and dividing it by the FixedU128 type
+        // precision. We first multiply the inner value to not end up saturating to 0 when dividing.
+        // Example: if treasury cut ends up being 5%, the inner value of the FixedU128 will be 50_000_000_000_000_000. If the
+        // amount to charge was 10_000_000_000 (one DOT), then the result of this will correctly be 500_000_000 (0.05 DOT).
+        amount_to_charge * treasury_cut.into_inner().into() / FixedU128::DIV.into()
+    }
+}
+
+/// Compute the fraction of charged tokens that go to the treasury using function
+///
+/// ```ignore
+/// I(x) = for x between 0 and x_ideal: 1 - x / x_ideal,
+/// for x between x_ideal and 1: 1 - 2^((x_ideal - x) / d)
+/// ```
+///
+/// where:
+/// * x is the system utilization rate, i.e. fraction of the total storage being provided in the system
+/// that is currently being used.
+/// * d is the falloff or `decay_rate`
+/// * x_ideal: the ideal system utilization rate.
+///
+/// The result is meant to be scaled with the minimum percentage amount and maximum percentage amount
+///  that goes to the treasury.
+///
+/// Arguments are:
+/// * `system_utilization`: The fraction of the total storage being provided in the system that is
+/// currently being used. Known as `x` in the literature. Must be between 0 and 1.
+/// * `ideal_system_utilization`: The fraction of total storage being provided in the system that should
+/// be being actively used. Known as `x_ideal` in the literature. Must be between 0 and 1.
+/// * `falloff`: Known as `decay_rate` in the literature. A co-efficient dictating the strength of
+///   the global incentivization to get the `ideal_system_utilization`. A higher number results in less typical
+///   funds to the treasury at the cost of greater volatility for providers. Must be more than 0.01.
+pub fn compute_percentage_to_treasury<P: PerThing>(
+    system_utilization: P,
+    ideal_system_utilization: P,
+    falloff: P,
+) -> P {
+    if system_utilization < ideal_system_utilization {
+        // ideal_system_utilization is more than 0 because it is strictly more than system_utilization
+        return (system_utilization / ideal_system_utilization).left_from_one();
+    }
+
+    if falloff < P::from_percent(1.into()) {
+        log::error!("Invalid inflation computation: falloff less than 1% is not supported");
+        return PerThing::zero();
+    }
```

Comment:
This limitation should be configured in the Config.

Reply:
I'll probably need help with this since I couldn't find in the docs how to limit a configuration parameter to a range of values